File: examples/go/_mock.go
//go:build exclude

// Phase 13.0 reference mock backend for Go SDK porters (Phase 13.6 build target).
// Carries //go:build exclude -- this file does NOT participate in any binary
// build during 13.0. Demo files (adtech.go, fraud.go, ecommerce.go) inline
// their own mockApp struct so each demo is a self-contained `go run` target.
//
// Per WARNING 8 fix (Option A) -- all 4 Go files declare package main; this
// _mock.go carries the build-exclude directive so it stays as reference code
// only. Trade-off: ~30 LOC duplication across the 3 demos; gain: zero-config
// `go run examples/go/<demo>.go`.
//
// Replaced by real beava-go client in Phase 13.6.
//
// Per Q2 locked answer + BLOCKER 4 checker fix: this mock COMPUTES features
// by applying registered descriptors on push. Demos go through the full
// register -> push -> query flow (no pre-seeding) so contract drift between
// specs and the real engine surfaces immediately at 13.6 re-verification.
//
// Supported ops in this mock (minimum for the 9 vertical demos):
//   - count: increment per matching event
//   - sum: accumulate field value
//   - mean: running sum / count
//   - min, max: comparison
//
// Sketches (NUnique, Quantile, TopK), decays, velocity, and geo ops are
// NOT computed -- demo files document the no-op fallback inline.

package main

import (
	"context"
	"fmt"
	"strings"
)

// AggSpec records one registered aggregation op.
type AggSpec struct {
	Op    string
	Field string // "" if no field (count)
}

// Descriptor is a registered event source or table aggregation.
type Descriptor struct {
	Name    string
	Kind    string // "event" | "table"
	Source  string // source event name (empty for kind="event")
	KeyCols []string
	Ops     map[string]AggSpec
}

// MockApp is the reference Phase 13.0 in-memory shim.
// Demo files inline an equivalent mockApp; this struct exists only as
// the canonical reference for Phase 13.6 SDK porters.
type MockApp struct {
	registered      []Descriptor
	tables          map[string]map[string]map[string]any
	aggState        map[string]map[string]float64
	registryVersion int
}

func NewMockApp() *MockApp {
	return &MockApp{
		registered: nil,
		tables:     make(map[string]map[string]map[string]any),
		aggState:   make(map[string]map[string]float64),
	}
}

// Register adds descriptors to the in-memory registry.
func (a *MockApp) Register(ctx context.Context, descs []Descriptor) (map[string]any, error) {
	a.registered = append(a.registered, descs...)
	a.registryVersion = len(a.registered)
	added := make([]string, 0, len(a.registered))
	for _, d := range a.registered {
		added = append(added, d.Name)
	}
	return map[string]any{
		"status":           "ok",
		"registry_version": a.registryVersion,
		"added":            added,
	}, nil
}

// Push applies registered table descriptors whose source matches the event.
func (a *MockApp) Push(ctx context.Context, eventName string, fields map[string]any) (map[string]any, error) {
	for _, desc := range a.registered {
		if desc.Kind != "table" {
			continue
		}
		if desc.Source != eventName {
			continue
		}
		key := keyFromEvent(desc.KeyCols, fields)
		for featureName, agg := range desc.Ops {
			a.update(desc.Name, key, featureName, agg, fields)
		}
	}
	return map[string]any{
		"ack_lsn":          1,
		"registry_version": a.registryVersion,
	}, nil
}

// Get returns the row-shape feature dict for (table, key).
func (a *MockApp) Get(ctx context.Context, table string, key string) (map[string]any, error) {
	if t, ok := a.tables[table]; ok {
		if row, ok := t[key]; ok {
			return row, nil
		}
	}
	return map[string]any{}, nil
}

// BatchGet fetches multiple (table, key) rows.
func (a *MockApp) BatchGet(ctx context.Context, requests []GetRequest) ([]map[string]any, error) {
	out := make([]map[string]any, 0, len(requests))
	for _, r := range requests {
		row, _ := a.Get(ctx, r.Table, r.Key)
		out = append(out, row)
	}
	return out, nil
}

// Reset clears all in-memory state.
func (a *MockApp) Reset(ctx context.Context) error {
	a.tables = make(map[string]map[string]map[string]any)
	a.aggState = make(map[string]map[string]float64)
	return nil
}

// Ping returns mock server info.
func (a *MockApp) Ping(ctx context.Context) (map[string]any, error) {
	return map[string]any{
		"server_version":   "0.0.0-mock",
		"registry_version": a.registryVersion,
	}, nil
}

// Close is a no-op for the mock.
func (a *MockApp) Close(ctx context.Context) error {
	return nil
}

// GetRequest is the batch_get request shape.
type GetRequest struct {
	Table string
	Key   string
}

func (a *MockApp) update(table, key, feature string, agg AggSpec, event map[string]any) {
	stateKey := table + "|" + key + "|" + feature
	state, ok := a.aggState[stateKey]
	if !ok {
		state = make(map[string]float64)
		a.aggState[stateKey] = state
	}
	switch agg.Op {
	case "count":
		state["count"]++
		a.setValue(table, key, feature, state["count"])
	case "sum":
		state["sum"] += getFloat(event, agg.Field)
		a.setValue(table, key, feature, state["sum"])
	case "mean":
		state["sum"] += getFloat(event, agg.Field)
		state["count"]++
		a.setValue(table, key, feature, state["sum"]/state["count"])
	case "min":
		v := getFloat(event, agg.Field)
		if _, seen := state["min_seen"]; !seen {
			state["min"] = v
			state["min_seen"] = 1
		} else if v < state["min"] {
			state["min"] = v
		}
		a.setValue(table, key, feature, state["min"])
	case "max":
		v := getFloat(event, agg.Field)
		if _, seen := state["max_seen"]; !seen {
			state["max"] = v
			state["max_seen"] = 1
		} else if v > state["max"] {
			state["max"] = v
		}
		a.setValue(table, key, feature, state["max"])
	default:
		// Unsupported in mock (sketches, decays, geo): no-op.
	}
}

func (a *MockApp) setValue(table, key, feature string, value any) {
	if _, ok := a.tables[table]; !ok {
		a.tables[table] = make(map[string]map[string]any)
	}
	if _, ok := a.tables[table][key]; !ok {
		a.tables[table][key] = make(map[string]any)
	}
	a.tables[table][key][feature] = value
}

func keyFromEvent(keyCols []string, event map[string]any) string {
	if len(keyCols) == 0 {
		return "_global"
	}
	parts := make([]string, len(keyCols))
	for i, k := range keyCols {
		parts[i] = fmt.Sprintf("%v", event[k])
	}
	return strings.Join(parts, "|")
}

func getFloat(event map[string]any, field string) float64 {
	if field == "" {
		return 0
	}
	v, ok := event[field]
	if !ok || v == nil {
		return 0
	}
	switch x := v.(type) {
	case float64:
		return x
	case float32:
		return float64(x)
	case int:
		return float64(x)
	case int64:
		return float64(x)
	}
	return 0
}

File: examples/typescript/_mock.ts
/**
 * Phase 13.0 mock backend for runnable TypeScript demos.
 *
 * In-memory App shim with MINIMAL AGGREGATION LOGIC. Drop-in for the
 * real `BeavaApp` during Phase 13.0 verification. Replaced by the real
 * `@beava/sdk` BeavaApp once Phase 13.6 lands (single-line edit per demo).
 *
 * Per Q2 locked answer + BLOCKER 4 checker fix: this mock COMPUTES
 * features by applying registered descriptors on push. Demos go through
 * the full register -> push -> query flow (no pre-seeding) so contract
 * drift between specs and the real engine surfaces immediately at
 * 13.6 re-verification.
 *
 * Supported ops in this mock (minimum for the 9 vertical demos):
 * - count: increment per matching event
 * - sum: accumulate field value
 * - mean: running sum / count
 * - min, max: comparison
 *
 * Sketches (nUnique, quantile, topK), decays, velocity, and geo ops
 * are NOT computed -- demo files document the no-op fallback inline.
 */

// camelCase API per docs/sdk-api/shared.md
// (Wire JSON keys are snake_case; the SDK transport layer would translate.)

type AggSpec = { op: string; field: string | null };

type Descriptor = {
  name: string;
  kind: "event" | "table";
  source: string | null;
  keyCols: string[];
  ops: Record<string, AggSpec>;
};

export class MockBeavaApp {
  private registered: Descriptor[] = [];
  private tables: Map<string, Map<string, Record<string, unknown>>> = new Map();
  private aggState: Map<string, Record<string, number>> = new Map();
  private registryVersion = 0;

  async register(
    descriptors: Descriptor[],
    opts?: { force?: boolean; dryRun?: boolean },
  ): Promise<{
    status: string;
    registryVersion: number;
    added: string[];
  }> {
    if (opts?.dryRun) {
      return {
        status: "ok",
        registryVersion: this.registryVersion,
        added: [],
      };
    }
    for (const d of descriptors) {
      this.registered.push(d);
    }
    this.registryVersion = this.registered.length;
    return {
      status: "ok",
      registryVersion: this.registryVersion,
      added: this.registered.map((d) => d.name),
    };
  }

  async push(
    eventName: string,
    fields: Record<string, unknown>,
  ): Promise<{ ackLsn: number; registryVersion: number }> {
    for (const desc of this.registered) {
      if (desc.kind !== "table") continue;
      if (desc.source !== eventName) continue;
      const key = this.keyFromEvent(desc.keyCols, fields);
      for (const [featureName, agg] of Object.entries(desc.ops)) {
        this.update(desc.name, key, featureName, agg, fields);
      }
    }
    return { ackLsn: 1, registryVersion: this.registryVersion };
  }

  async get(
    table: string,
    key: string | (string | number | boolean)[],
  ): Promise<Record<string, unknown>> {
    const keyStr = Array.isArray(key) ? key.map(String).join("|") : key;
    return this.tables.get(table)?.get(keyStr) ?? {};
  }

  async batchGet(
    requests: Array<{
      table: string;
      key: string | (string | number | boolean)[];
    }>,
  ): Promise<Record<string, unknown>[]> {
    const out: Record<string, unknown>[] = [];
    for (const r of requests) {
      out.push(await this.get(r.table, r.key));
    }
    return out;
  }

  async reset(): Promise<void> {
    this.tables.clear();
    this.aggState.clear();
  }

  async ping(): Promise<{ serverVersion: string; registryVersion: number }> {
    return {
      serverVersion: "0.0.0-mock",
      registryVersion: this.registryVersion,
    };
  }

  async close(): Promise<void> {
    // no-op
  }

  private keyFromEvent(
    keyCols: string[],
    event: Record<string, unknown>,
  ): string {
    if (keyCols.length === 0) return "_global";
    return keyCols.map((k) => String(event[k] ?? "")).join("|");
  }

  private update(
    table: string,
    key: string,
    feature: string,
    agg: AggSpec,
    event: Record<string, unknown>,
  ): void {
    const stateKey = `${table}|${key}|${feature}`;
    const state = this.aggState.get(stateKey) ?? {};
    const op = agg.op;
    if (op === "count") {
      state.count = (state.count ?? 0) + 1;
      this.setValue(table, key, feature, state.count);
    } else if (op === "sum") {
      const v = agg.field ? Number(event[agg.field] ?? 0) : 0;
      state.sum = (state.sum ?? 0) + v;
      this.setValue(table, key, feature, state.sum);
    } else if (op === "mean") {
      const v = agg.field ? Number(event[agg.field] ?? 0) : 0;
      state.sum = (state.sum ?? 0) + v;
      state.count = (state.count ?? 0) + 1;
      this.setValue(table, key, feature, state.sum / state.count);
    } else if (op === "min") {
      const v = agg.field ? Number(event[agg.field] ?? 0) : 0;
      state.min = state.min === undefined ? v : Math.min(state.min, v);
      this.setValue(table, key, feature, state.min);
    } else if (op === "max") {
      const v = agg.field ? Number(event[agg.field] ?? 0) : 0;
      state.max = state.max === undefined ? v : Math.max(state.max, v);
      this.setValue(table, key, feature, state.max);
    }
    // Unsupported ops in mock (sketches, decays, geo, etc.): no-op.
    this.aggState.set(stateKey, state);
  }

  private setValue(
    table: string,
    key: string,
    feature: string,
    value: unknown,
  ): void {
    if (!this.tables.has(table)) this.tables.set(table, new Map());
    const t = this.tables.get(table)!;
    if (!t.has(key)) t.set(key, {});
    t.get(key)![feature] = value;
  }
}

// Drop-in factory: `const app = new BeavaApp();` -> swap to real `@beava/sdk`
// in Phase 13.6 by replacing this import.
export class BeavaApp extends MockBeavaApp {}

// Demo helpers -- describe descriptors inline, in real Beava these would
// be `bv.event({...})` / `bv.table({...})` builder calls.

export function event(name: string): Descriptor {
  return { name, kind: "event", source: null, keyCols: [], ops: {} };
}

export function table(args: {
  name: string;
  source: string;
  key: string | string[];
  ops: Record<string, [string, string | null]>;
}): Descriptor {
  const keyCols = typeof args.key === "string" ? [args.key] : args.key;
  const ops: Record<string, AggSpec> = {};
  for (const [feat, [op, field]] of Object.entries(args.ops)) {
    ops[feat] = { op, field };
  }
  return {
    name: args.name,
    kind: "table",
    source: args.source,
    keyCols,
    ops,
  };
}
